### ISSUE TITLE
SchemaOf incompatible with typescript index signature

### ISSUE THREAD

**Issue by olivervorasai** [jquense/yup#1510]
**Describe the bug**
When using `SchemaOf` to give a type to a validation schema, typescript gives an error of `Type 'Reference<unknown>' is missing the following properties from type 'BaseSchema<Maybe<string>, AnyObject, string>': type, __inputType, __outputType, __isYupSchema__, and 45 more.ts(2322)`

```typescript
import * as yup from "yup";

interface Vegetable {
  [key: string]: string; // Not supported?
  name: string;
}

export const validationSchema: yup.SchemaOf<Vegetable> = yup.object({
  name: yup.string().required()
});
```

**To Reproduce**

https://codesandbox.io/s/vjprr?file=/src/index.ts

**Expected behavior**
There should be no typescript errors.

**Platform (please complete the following information):**
 - Browser Chrome
 - Version 95.0.4638.69 (Official Build) (64-bit)

**Additional context**
This is the full error returned by the editor

`Type 'OptionalObjectSchema<{ name: RequiredStringSchema<string | undefined, AnyObject>; }, AnyObject, TypeOfShape<{ name: RequiredStringSchema<string | undefined, AnyObject>; }>>' is not assignable to type 'ObjectSchemaOf<Vegetable, never>'.
  The types returned by 'concat(...)' are incompatible between these types.
    Type 'ObjectSchema<{ name: RequiredStringSchema<string | undefined, AnyObject>; } & ObjectShape, AnyObject, TypeOfShape<{ name: RequiredStringSchema<string | undefined, AnyObject>; } & ObjectShape> | null | undefined, AssertsShape<...> | ... 1 more ... | undefined>' is not assignable to type 'ObjectSchema<{ [x: string]: BaseSchema<Maybe<string>, AnyObject, string>; name: BaseSchema<Maybe<string>, AnyObject, string>; } & ObjectShape, AnyObject, TypeOfShape<...> | ... 1 more ... | undefined, AssertsShape<...> | ... 1 more ... | undefined>'.
      Type '{ name: RequiredStringSchema<string | undefined, AnyObject>; } & ObjectShape' is not assignable to type '{ [x: string]: BaseSchema<Maybe<string>, AnyObject, string>; name: BaseSchema<Maybe<string>, AnyObject, string>; } & ObjectShape'.
        Type '{ name: RequiredStringSchema<string | undefined, AnyObject>; } & ObjectShape' is not assignable to type '{ [x: string]: BaseSchema<Maybe<string>, AnyObject, string>; name: BaseSchema<Maybe<string>, AnyObject, string>; }'.
          'string' index signatures are incompatible.
            Type 'AnySchema<any, any, any> | Reference<unknown> | Lazy<any, any>' is not assignable to type 'BaseSchema<Maybe<string>, AnyObject, string>'.
              Type 'Reference<unknown>' is missing the following properties from type 'BaseSchema<Maybe<string>, AnyObject, string>': type, __inputType, __outputType, __isYupSchema__, and 45 more.ts(2322)`

**Comment by natech3ng** [comment]
It's good if I downgrade it to 0.29, but not 0.32

**Comment by olivervorasai** [comment]
This is fixed in `1.0.0-beta.1`
